### PR TITLE
Make offline compression and decompression independent of STATIC_URL

### PR DESCRIPTION
**Pull request by kottenator**
If `settings.COMPRESS_OFFLINE_URL_PLACEHOLDERS` is `True` then:
1. Replace `settings.COMPRESS_URL` with `settings.COMPRESS_URL_PLACEHOLDER` in the offline manifest when doing offline compression via `manage.py compress`
2. Calculate offline manifest keys with `settings.STATIC_URL`replaced with `''`

In this way we can do offline compression with one `settings.STATIC_URL` and then deploy it with different `settings.STATIC_URL` - `manifest.json` will work with any static URL.

More details are in #778.

### CODE REVIEW

**Comment by karyon** [comment]
unfortunately i don't know enough of the code to see whether there might be a better way. right now this feels rather bolted on top. @kottenator have you thought about any other ways? @mat any thoughts on this?

**Comment by diox** [comment]
Sorry, I don't really like the approach. Altering settings at run-time is a bad idea IMHO, and doing it for a pretty uncommon use-case doesn't seem worth it.

**Comment by kottenator** [comment]
The problem is in Compressor approach - it renders the node contents, which often are `{% static %}` tags, which use `settings.STATIC_URL`. 

There is no better way to implement the URL independence.

**Comment by kottenator** [comment]
And also - I've made this optional, and only for offline compression.

**Comment by kottenator** [comment]
Well, actually there may be a way without modifying `settings`, but it would be something like:

* render `{% compress %}` node
* replace all `STATIC_URL` entries with placeholder
* calculate the manifest key

**Comment by kottenator** [comment]
It was easy to change :)
@diox, @karyon - please, take a look again.

File: compressor/management/commands/compress.py [hunk 1]
```diff
@@ -250,7 +250,15 @@ def compress(self, log=None, **options):
...
-                    offline_manifest[key] = result
+
+                    if settings.COMPRESS_OFFLINE_USE_URL_PLACEHOLDER:
+                        # Replace settings.COMPRESS_URL with a placeholder
+                        offline_manifest[key] = result.replace(
+                            settings.COMPRESS_URL, settings.COMPRESS_URL_PLACEHOLDER
+                        )
+                    else:
+                        offline_manifest[key] = result
+
```
Comment (karyon) [inline compressor/management/commands/compress.py]: i was wondering, why don't you do `result = result.replace(....)` as you do in `templatetags/compress.py `? this makes this method return results that did not get written to the manifest.

looking at this i noticed that this `compress` method writes the manifest, and the `handle` method reads that manifest, merges it with others, and writes again. wouldn't it be better to not write the manifest here, just return it, and let `handle` do the writing? @mat?

File: compressor/tests/test_offline_url_placeholders.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+for key in dir(compressor.tests.test_offline):
+    value = getattr(compressor.tests.test_offline, key)
+    if (
+        key not in ('OfflineCompressStaticTemplateTagTestCase', 'TestCompressCommand') and
+        # Find all the TestCase subclasess except the TestCase itself
+        isinstance(value, type) and issubclass(value, TestCase) and value is not TestCase
+    ):
+        new_class = type(key, (value,), {})
+        module_dict[key] = override_settings(COMPRESS_OFFLINE_USE_URL_PLACEHOLDER=True)(new_class)
```
Comment (karyon) [inline compressor/tests/test_offline_url_placeholders.py]: what does all this do?

File: compressor/contrib/jinja2ext.py [hunk 3]
```diff
@@ -45,11 +45,10 @@ def _compress_normal(self, kind, mode, caller):
...
-        original_content = caller()
         context = {
-            'original_content': original_content
+            'original_content_generator': caller,
         }
         return self.render_compressed(context, kind, mode, forced=forced)
 
     def get_original_content(self, context):
-        return context['original_content']
+        return context['original_content_generator']()
```
Comment (karyon) [inline compressor/contrib/jinja2ext.py]: these seem unrelated...?

File: compressor/management/commands/compress.py (hunk 1 above)
Comment (karyon) [inline compressor/management/commands/compress.py]: sorry, @diox :)

**Comment by karyon** [comment]
added some random comments, didn't do an actual review. i'd wait for @diox opinion on the general approach (and the offline manifest writing thing i commented on)

File: compressor/management/commands/compress.py (hunk 1 above)
Comment (kottenator) [inline compressor/management/commands/compress.py]: Yes, I see what you mean: writing the manifest, then reading, then writing again. 
I'm not sure though if it'll give the same effect if I'll move it to `.handle()`...

File: compressor/tests/test_offline_url_placeholders.py (hunk 2 above)
Comment (kottenator) [inline compressor/tests/test_offline_url_placeholders.py]: There is `compress/tests/test_offline.py` file, it has ~60 tests. I want to re-use them: to run them all but with `settings.COMPRESS_OFFLINE_USE_URL_PLACEHOLDER=True`.

I'm dynamically creating subclasses with overridden setting.

File: compressor/contrib/jinja2ext.py (hunk 3 above)
Comment (kottenator) [inline compressor/contrib/jinja2ext.py]: I wanted to ask you about this.

In Django you always re-render the node contents but in Jinja you render it once and then re-use. This has brought some problems when I was overriding Django settings. Now it doesn't, but shouldn't it be implemented in the same way for Django and Jinja?

Comment (karyon) [inline compressor/contrib/jinja2ext.py]: i don't know :) but @diox might...? if you can write a test that demonstrates the problem, i could merge this, but i'd also like to have this in a separate commit.

File: compressor/tests/test_offline_url_placeholders.py (hunk 2 above)
Comment (karyon) [inline compressor/tests/test_offline_url_placeholders.py]: okay... provided @diox signs off the general approach of this PR, this part here definitely needs a comment explaining exactly that.

File: compressor/contrib/jinja2ext.py (hunk 3 above)
Comment (kottenator) [inline compressor/contrib/jinja2ext.py]: Yes, probably it's a separate thing. I'll remove it now.

Comment (kottenator) [inline compressor/contrib/jinja2ext.py]: Reverted.

File: compressor/management/commands/compress.py (hunk 1 above)
Comment (karyon) [inline compressor/management/commands/compress.py]: for the record the double write was added in https://github.com/django-compressor/django-compressor/pull/806

**Comment by kottenator** [comment]
@diox - I've made all (except one) the change requests and have removed `settings` modification. Please, review this PR again, when you'll have a chance.

**Comment by diox** [comment]
I haven't got the time to do a proper review, but I like this approach a lot better. The manipulation to run all the tests is kinda hairy, but I suppose you don't have much choice with the way they are written (if I could do it all over again I'd do a single class).

We'll need docs about the new settings.

**Comment by kottenator** [comment]
Sure, I'll add the docs.

**Comment by kottenator** [comment]
Docs added.

File: compressor/management/commands/compress.py (hunk 1 above)
Comment (karyon) [inline compressor/management/commands/compress.py]: huh, even your own test uses the return values. i don't get it. i hope you can enlighten me :)

File: docs/settings.txt [hunk 4]
```diff
@@ -516,3 +516,21 @@ Offline settings
...
+
+.. attribute:: COMPRESS_OFFLINE_URLLESS
+
+    :Default: ``False``
+
+    Enable URL-less mode when you do offline compression & decompression regardless of
+    ``settings.STATIC_URL`` and ``settings.COMPRESS_URL`` so your assets become re-locatable
+    for any URLs.
+
+    Affects manifest file contents and manifest key determination, but not the assets contents.
```
Comment (karyon) [inline docs/settings.txt]: "for any URL" and "asset's"

File: docs/settings.txt [hunk 5]
```diff
@@ -516,3 +516,21 @@ Offline settings
...
+
+.. attribute:: COMPRESS_URL_PLACEHOLDER
+
+    :Default: ``'/__compressor_url_placeholder__/'``
+
+    If :attr:`~django.conf.settings.COMPRESS_OFFLINE_URLLESS` is ``True`` -
+    replace :attr:`~django.conf.settings.COMPRESS_URL` with this placeholder in the
+    offline manifest file.
```
Comment (karyon) [inline docs/settings.txt]:
```
    If :attr:`~django.conf.settings.COMPRESS_OFFLINE_URLLESS` is ``True``,
    each occurrence of :attr:`~django.conf.settings.COMPRESS_URL` in the offline
    manifest file will be replaced with this placeholder.
```

File: compressor/management/commands/compress.py [hunk 6]
```diff
@@ -250,7 +250,15 @@ def compress(self, log=None, **options):
                     except Exception as e:
                         raise CommandError("An error occurred during rendering %s: "
                                            "%s" % (template.template_name, smart_text(e)))
-                    offline_manifest[key] = result
+
+                    if settings.COMPRESS_OFFLINE_URLLESS:
+                        # Replace settings.COMPRESS_URL with a placeholder
```
Comment (karyon) [inline compressor/management/commands/compress.py]: as below, remove comment

File: docs/settings.txt (hunk 5 above)
Comment (kottenator) [inline docs/settings.txt]: Done.

File: docs/settings.txt (hunk 4 above)
Comment (kottenator) [inline docs/settings.txt]: Done.

File: compressor/templatetags/compress.py [hunk 7]
```diff
@@ -64,7 +64,15 @@ def render_offline(self, context):
         key = get_offline_hexdigest(self.get_original_content(context))
         offline_manifest = get_offline_manifest()
         if key in offline_manifest:
-            return offline_manifest[key]
+            result = offline_manifest[key]
+
+            if settings.COMPRESS_OFFLINE_URLLESS:
+                # Replace placeholder with original settings.COMPRESS_URL
```
Comment (kottenator) [inline compressor/templatetags/compress.py]: Done.

File: compressor/conf.py [hunk 8]
```diff
@@ -70,6 +70,12 @@ class CompressorConf(AppConf):
     OFFLINE_MANIFEST = 'manifest.json'
     # The Context to be used when TemplateFilter is used
     TEMPLATE_FILTER_CONTEXT = {}
+    # URL-less mode: you do offline compression & decompression regardless of settings.STATIC_URL
+    # and settings.COMPRESS_URL so your assets become re-locatable for any URLs.
+    # Affects manifest file contents and manifest key determination, but not the assets contents.
+    OFFLINE_URLLESS = False
+    # Placeholder to be used instead of settings.COMPRESS_URL while offline compression
```
Comment (kottenator) [inline compressor/conf.py]: Done.

File: compressor/management/commands/compress.py (hunk 1 above)
Comment (kottenator) [inline compressor/management/commands/compress.py]: _I've replied but now I've lost the reply, so I'll reply again_ :)

I agree, that's a controversial part. My idea was:

* `Command.compress()` returns full functional HTML code so you can use it
* when `manifest.json` contains _unprocessed_ (technical) code that should be _processed first_ before usage

If it's too controversial - I could try to simplify it.

File: compressor/management/commands/compress.py (hunk 6 above)
Comment (kottenator) [inline compressor/management/commands/compress.py]: Done.

Comment (kottenator) [inline compressor/management/commands/compress.py]: Done.

File: compressor/cache.py [hunk 9]
```diff
@@ -50,6 +50,9 @@ def get_mtime_cachekey(filename):
 
 
 def get_offline_hexdigest(render_template_string):
+    if settings.COMPRESS_OFFLINE_URLLESS:
+        # Make hexdigest calculation independent of settings.STATIC_URL
+        render_template_string = render_template_string.replace(settings.STATIC_URL, '')
```
Comment (kottenator) [inline compressor/cache.py]: BTW, maybe not in this PR, but how do you like the idea - to always do this?

File: compressor/conf.py [hunk 10]
```diff
@@ -70,6 +70,12 @@ class CompressorConf(AppConf):
     OFFLINE_MANIFEST = 'manifest.json'
     # The Context to be used when TemplateFilter is used
     TEMPLATE_FILTER_CONTEXT = {}
+    # URL-less mode: you do offline compression & decompression regardless of settings.STATIC_URL
+    # and settings.COMPRESS_URL so your assets become re-locatable for any URLs.
+    # Affects manifest file contents and manifest key determination, but not the assets contents.
+    OFFLINE_URLLESS = False
```
Comment (karyon) [inline compressor/conf.py]: in my opinion the key determination is a technical detail that's done purely to make the placeholder thing work.... @diox any thoughts on this?

File: compressor/management/commands/compress.py (hunk 1 above)
Comment (kottenator) [inline compressor/management/commands/compress.py]: Let me try to simplify it then...

File: compressor/conf.py (hunk 10 above)
Comment (kottenator) [inline compressor/conf.py]: IMO, the placeholder is a technical detail of `STATIC_URL`-independent mode implementation :) 

Nobody should care how is it implemented: using placeholders or by generating manifest in absolutely different way.

File: compressor/management/commands/compress.py (hunk 1 above)
Comment (kottenator) [inline compressor/management/commands/compress.py]: Ok, I've done what you had asked. Now `Command.compress()` returns exactly what is stored in the manifest.

I had to refactor `test_offline.py` to make it _reusable_ for this case.

File: compressor/conf.py (hunk 10 above)
Comment (diox) [inline compressor/conf.py]: My 2 cents:
- `OFFLINE_URLLESS=True` should just be the default behavior and not exposed as a setting.
- `URL_PLACEHOLDER` doesn't technically need to be exposed as a setting.

Comment (karyon) [inline compressor/conf.py]: @diox just to make sure i get this right, you propose to remove the URLLESS setting altogether and just always do the replacement with no option to turn it off? as in, if we had seen earlier that this approach is cool, most of @kottenator's work wouldn't have been necessary ;) ? (i think we should do it this way, just am a bit annoyed that i/we didn't think of it earlier)

@kottenator can you think of any downsides?

Comment (karyon) [inline compressor/conf.py]: i'd say in this case `URL_PLACEHOLDER` can still be a setting because that's a good place for global constants (or is there a good alternative?), but not document it because people should generally not care about it.

Comment (diox) [inline compressor/conf.py]: Yes. Sorry for the back-and-forth, but I think it's worth it to reduce complexity. For the record I still think the original issue is not a very common use-case, but after thinking about this for a while I don't see much downside in always doing the replace.

Comment (diox) [inline compressor/conf.py]: > i'd say in this case URL_PLACEHOLDER can still be a setting because that's a good place for global constants (or is there a good alternative?), but not document it because people should generally not care about it.

Fair enough.

File: compressor/cache.py [hunk 11]
```diff
@@ -50,7 +50,7 @@ def get_mtime_cachekey(filename):
 
 
 def get_offline_hexdigest(render_template_string):
-    return get_hexdigest(render_template_string)
+    return get_hexdigest(render_template_string.replace(settings.STATIC_URL, ''))
```
Comment (diox) [inline compressor/cache.py]: Add a comment explaining why

File: compressor/tests/test_offline.py [hunk 12]
```diff
@@ -72,10 +72,8 @@ def setUp(self):
         # TEMPLATES[1] to be Jinja2 templates backend in test_settings.
         TEMPLATES = copy.deepcopy(settings.TEMPLATES)
 
-        django_template_dir = os.path.join(
-            TEMPLATES[0]['DIRS'][0], self.templates_dir)
-        jinja2_template_dir = os.path.join(
-            TEMPLATES[1]['DIRS'][0], self.templates_dir)
+        django_template_dir = os.path.join(TEMPLATES[0]['DIRS'][0], self.templates_dir)
+        jinja2_template_dir = os.path.join(TEMPLATES[1]['DIRS'][0], self.templates_dir)
```
Comment (diox) [inline compressor/tests/test_offline.py]: Why change this ? It's now breaking 79 chars limit, which was the reason for the current wrapping (I know all compressor code doesn't follow pep8, but let's not regress on the files that do)

Comment (diox) [inline compressor/tests/test_offline.py]: Same for a lot of changes below.

Comment (kottenator) [inline compressor/tests/test_offline.py]: Sure, I'll bring it back.

File: compressor/cache.py (hunk 11 above)
Comment (kottenator) [inline compressor/cache.py]: OK, will do now.

File: compressor/tests/test_offline.py (hunk 12 above)
Comment (kottenator) [inline compressor/tests/test_offline.py]: I'm just using 99 chars limit, which is [allowed by PEP8](https://www.python.org/dev/peps/pep-0008/#maximum-line-length). But I don't want to break your rules.

File: compressor/management/commands/compress.py [hunk 13]
```diff
@@ -246,7 +246,9 @@ def compress(self, log=None, **options):
                         continue
 
                     try:
-                        result = parser.render_node(template, context, node)
+                        result = parser.render_node(template, context, node).replace(
+                            settings.COMPRESS_URL, settings.COMPRESS_URL_PLACEHOLDER
```
Comment (karyon) [inline compressor/management/commands/compress.py]: please do the replace outside the try block

File: compressor/tests/test_offline.py [hunk 14]
```diff
@@ -364,6 +357,18 @@ class OfflineCompressStaticTemplateTagTestCase(OfflineTestCaseMixin, TestCase):
...
+        count, result = CompressCommand().compress(
+            log=self.log, verbosity=self.verbosity, engine=engine
+        )
+        self.assertEqual(1, count)
+        self.assertEqual([self._render_script(self.expected_hash)], result)
+        self.assertEqual(self._render_template(engine), self._render_result(result))
+
+        # Changing settings.STATIC_URL doesn't affect the next "offline" decompression
+        with self.settings(STATIC_URL='/another/static/url/'):
+            self.assertEqual(self._render_template(engine), self._render_result(result))
```
Comment (karyon) [inline compressor/tests/test_offline.py]: could you make it more explicit what is being tested here? i don't fully understand the structure of the tests but my feeling is this should go in a new Class whose name roughly describes what's being tested. 

also, could you clarify what you mean by "decompression" in the comment? we currently don't use that word anywhere

File: compressor/tests/test_offline.py (hunk 12 above)
Comment (kottenator) [inline compressor/tests/test_offline.py]: Done, 79 chars limit is back to this file (and even better than it was).

File: compressor/tests/test_offline.py [hunk 15]
```diff
@@ -364,6 +376,21 @@ class OfflineCompressStaticTemplateTagTestCase(OfflineTestCaseMixin, TestCase):
...
+        self.assertEqual(1, count)
+        self.assertEqual([self._render_script(self.expected_hash)], result)
+        self.assertEqual(
+            self._render_template(engine), self._render_result(result))
+
+        # Changing settings.STATIC_URL doesn't affect the next
+        # offline "decompression"
+        with self.settings(STATIC_URL='/another/static/url/'):
+            self.assertEqual(
+                self._render_template(engine), self._render_result(result))
```
Comment (karyon) [inline compressor/tests/test_offline.py]: that comment got hidden for some reason so here it is again:



could you make it more explicit what is being tested here? i don't fully understand the structure of the tests but my feeling is this should go in a new Class whose name roughly describes what's being tested.

also, could you clarify what you mean by "decompression" in the comment? we currently don't use that word anywhere

File: compressor/management/commands/compress.py (hunk 13 above)
Comment (kottenator) [inline compressor/management/commands/compress.py]: Done.

File: compressor/tests/test_offline.py (hunk 14 above)
Comment (kottenator) [inline compressor/tests/test_offline.py]: Sure, I'll do a separate test class.

By "offline decompression" I mean extraction & processing of the offline manifest content into the HTML output.

File: compressor/tests/test_offline.py (hunk 15 above)
Comment (kottenator) [inline compressor/tests/test_offline.py]: Sure, I'll do a separate test class.

By "offline decompression" I mean:

* offline manifest key determination
* extraction & processing of the offline manifest content into the HTML output

File: compressor/tests/test_offline.py [hunk 16]
```diff
@@ -471,6 +460,40 @@ class OfflineCompressTestCaseWithContextGeneratorSuper(
...
+
+    We use COMPRESS_OFFLINE_CONTEXT generator to make sure that
+    STATIC_URL is not cached when rendering the template.
+    """
+    templates_dir = 'test_static_url_independence'
+    expected_hash = '44ed960a3d1d'
+    additional_test_settings = {
+        'STATIC_URL': '/custom/static/url/',
+        'COMPRESS_OFFLINE_CONTEXT': (
+            'compressor.tests.test_offline.static_url_context_generator'
```
Comment (karyon) [inline compressor/tests/test_offline.py]: did you want to specifically use the offline context thing? if not, you could just use the static tag in the template and wouldn't need the offline context

Comment (karyon) [inline compressor/tests/test_offline.py]: ...okay i see. you can still use a dictionary here instead of the generator, that'll be a bit less code.

Comment (kottenator) [inline compressor/tests/test_offline.py]: The problem with `{% static %}` tag is that it has _cached_ storage.

OK, I'll try to use a dict.

Comment (kottenator) [inline compressor/tests/test_offline.py]: No, unfortunately `dict` is also getting cached because it's calculated only once, during tests _compilation_.

**Comment by karyon** [comment]
@kottenator you're awesome for being so patient! :) and thank you for your contribution.

@diox i saw your two review comments addressed and assumed there's nothing more. if yes, i can fix that.

**Comment by kottenator** [comment]
Thank you, @karyon and @diox! Keep your framework being awesome! ;)
